src/cryptonote_core/service_node_list.cpp

```diff
@@ -321,9 +329,214 @@ namespace service_nodes
       LOG_PRINT_L1("Deregistration for service node: " << key);
     }
 
-
     m_rollback_events.push_back(std::unique_ptr<rollback_event>(new rollback_change(block_height, key, iter->second)));
     m_service_nodes_infos.erase(iter);
+
+    return true;
+  }
+
+  static uint64_t get_new_swarm_id(std::mt19937_64& mt, const std::vector<swarm_id_t>& ids)
+  {
+    uint64_t id_new = 0;
+
+    while (id_new == 0 && ( std::find(ids.begin(), ids.end(), id_new) != ids.end() )) {
+      id_new = uniform_distribution_portable(mt, UINT64_MAX);
+    }
+
+    return id_new;
+  }
+
+  static std::vector<swarm_id_t> get_all_swarms(const std::map<swarm_id_t, std::vector<crypto::public_key>>& swarm_to_snodes)
+  {
+    std::vector<swarm_id_t> all_swarms;
+    all_swarms.reserve(swarm_to_snodes.size());
+    for (const auto& entry : swarm_to_snodes) {
+      all_swarms.push_back(entry.first);
+    }
+    return all_swarms;
+  }
+
+  static crypto::public_key pop_random_snode(std::mt19937_64& mt, std::vector<crypto::public_key>& vec)
+  {
+    const auto idx = uniform_distribution_portable(mt, vec.size());
+    const auto sn_pk = vec.at(idx);
+    auto it = vec.begin();
+    std::advance(it, idx);
+    vec.erase(it);
+    return sn_pk;
+  }
+
+
+  static void calc_swarm_changes(std::map<swarm_id_t, std::vector<crypto::public_key>>& swarm_to_snodes, uint64_t seed)
+  {
+    std::mt19937_64 mersenne_twister(seed);
+
+    std::vector<crypto::public_key> swarm_buffer = swarm_to_snodes[QUEUE_SWARM_ID];
+    swarm_to_snodes.erase(QUEUE_SWARM_ID);
+
+    auto all_swarms = get_all_swarms(swarm_to_snodes);
+    std::sort(all_swarms.begin(), all_swarms.end());
+      /// shuffle to remove all biases
+    loki_shuffle(all_swarms, seed);
+
+    /// prepare a lambda to be used with algorithms
```

Comment:
Some of the comments are superfluous and can be removed since the intent is clear in the code. The 1, 2, 3 steps are informative but this one and the ones at L419, 428, 442, 486 etc are noisy.

Reply:
They helped me structure my thoughts when writing the code, so hopefully they will be useful for someone in the future. There is always a gap between the intent and what the code effectively does, comments help clarify that.

I removed the comment about the lambda, that one was not useful.